File: Example/OpAdxSdkDemo_OC/OpAdxDemo_OC/Classes/AppOpenViewController.m
//
//  AppOpenViewController.m
//  OpAdxSdkDemo_OC
//
//  Created by Luan Chen on 2026/1/26.
//

#import "AppOpenViewController.h"
#import <OpAdxSdk/OpAdxSdk.h>

@interface AppOpenViewController ()<OpAdxAppOpenAdDelegate>

@property (nonatomic, strong, nullable) OpAdxAppOpenAdBridge *appOpenAd;

@end

@implementation AppOpenViewController

- (BOOL)hasVideo {
    return YES;
}

- (AdFormat)adFormat {
    return AdFormatAppOpen;
}

- (NSString *)adFormatString {
    return @"AppOpen Ad";
}

- (void)loadAd {
    if (!self.placementId) return;
    
    [self.logView print:@"Loading ..."];
    if (self.appOpenAd != nil) {
        [self destroyAd];
    }
    
    OpAdxAppOpenAdBridge *appOpenAd = [[OpAdxAppOpenAdBridge alloc] initWithPlacementId:self.placementId auctionType: AdAuctionTypeRegular];
    self.appOpenAd = appOpenAd;
    appOpenAd.delegate = self;
    [appOpenAd loadAd];
}

- (void)showAd {
    if (!self.appOpenAd) return;
    
    if (![self.appOpenAd isAdValid]) {
        [self.logView print:@"Ad is invalidated."];
        [self destroyAd];
        return;
    }
    
    __weak typeof(self) weakSelf = self;
    dispatch_after(dispatch_time(DISPATCH_TIME_NOW, (int64_t)(0.5 * NSEC_PER_SEC)), dispatch_get_main_queue(), ^{
        [weakSelf.appOpenAd showAdFrom:weakSelf];
    });
    
    [self enableDestroyAd];
    [self disableShowAd];
}

- (void)destroyAd {
    [super destroyAd];
    self.appOpenAd = nil;
    [self disableDestroyAd];
    [self disableShowAd];
}

#pragma mark - OpAdxAppOpenAdDelegate

- (void)appOpenAdDidLoad:(OpAdxAppOpenAdBridge *)appOpenAd {
    NSLog(@"[ADX] AppOpen广告加载成功: %@", appOpenAd.placementId);
    [self.logView print:[NSString stringWithFormat:@"Loaded."]];
   
    [self enableShowAd];
    [self enableDestroyAd];
    
}

- (void)appOpenAd:(OpAdxAppOpenAdBridge *)appOpenAd didFailWithError:(OpAdxAdError *)error {
    NSLog(@"[ADX] AppOpen广告加载失败: %@ %@", appOpenAd.placementId, error);
    [self.logView print:error.message];
    
}

- (void)appOpenAdDidClick:(OpAdxAppOpenAdBridge *)appOpenAd {
    NSLog(@"[ADX] AppOpen广告被点击: %@", appOpenAd.placementId);
    [self.logView print:@"onAdDidClick"];
    
}

- (void)appOpenAdDidClose:(OpAdxAppOpenAdBridge *)appOpenAd {
    NSLog(@"[ADX] AppOpen广告关闭: %@", appOpenAd.placementId);
    [self.logView print:@"onAdDidClose"];
    // 清理
    self.appOpenAd = nil;

}

- (void)appOpenAdWillLogImpression:(OpAdxAppOpenAdBridge *)appOpenAd {
    NSLog(@"[ADX] AppOpen广告展示: %@", appOpenAd.placementId);
    [self.logView print:@"onAdImpression"];
    
}

@end

File: Example/OpAdxSdkDemo_Swift/OpAdxSdkDemo/AppOpenViewController.swift
//
//  AppOpenViewController.swift
//  OpAdxSdkDemo
//
//  Created by Luan Chen on 2026/1/19.
//

import UIKit
import OpAdxSdk

class AppOpenViewController: BaseViewController {
    
    private var appOpenAd: OpAdxAppOpenAd?
    
    override var hasVideo: Bool {
        return false
    }
    
    override var adFormat: AdFormat {
        return .appOpen
    }
    
    override var adFormatString: String {
        return "App Open Ad"
    }
    
    override func loadAd() {
        guard let placementId = placementId else { return }
        
        logView.print("Loading ...")
        if appOpenAd != nil {
            self.destroyAd()
        }
        appOpenAd = OpAdxAppOpenAd(placementId: placementId, auctionType: AdAuctionType.regular)
        
        // 创建监听器实例
        let listener = OpAdxAppOpenAdLoadListenerImp(
            onAdLoaded: { [weak self] ad in
                guard let self = self, let ad = ad as? OpAdxAppOpenAd else { return }
                self.appOpenAd = ad
                self.logView.print("Loaded")
                self.enableShowAd()
                self.enableDestroyAd()
            },
            onAdFailedToLoad: { [weak self] error in
                guard let self = self else { return }
                self.logView.print(error.message)
            }
        )
        
        appOpenAd?.load(placementId: placementId, listener: listener)
    }
    
    override func showAd() {
        guard let appOpenAd = appOpenAd else { return }
        
        if appOpenAd.isAdInvalidated() {
            logView.print("Ad is invalidated.")
            destroyAd()
            return
        }
        
        // 创建交互监听器
        let interactionListener = OpAdxAppOpenAdInteractionListenerImp(
            onAdClicked: { [weak self] in
                guard let self = self else { return }
                self.logView.print("Clicked!")
            },
            onAdDisplayed: { [weak self] in
                guard let self = self else { return }
                self.logView.print("Displayed!")
            },
            onAdDismissed: { [weak self] in
                guard let self = self else { return }
                self.logView.print("Dismissed")
                self.destroyAd()
            },
            onAdFailedToShow: { [weak self] error in
                guard let self = self else { return }
                self.logView.print(error.message)
            }
        )
        
        appOpenAd.show(on: self, listener: interactionListener)
        
        enableDestroyAd()
        disableShowAd()
    }
    
    override func destroyAd() {
        super.destroyAd()
        appOpenAd?.destroy()
        appOpenAd = nil
        logView.print("Destroyed...")
        disableDestroyAd()
        disableShowAd()
    }
    
    deinit {
        destroyAd()
    }
}
